lib/spack/spack/util/windows_registry.py

```diff
@@ -218,51 +243,128 @@ def get_subkeys(self):
             raise RegistryError("Cannot query values from invalid key %s" % self.key)
         with self.invalid_reg_ref_error_handler():
             return self.reg.values
 
-    def _traverse_subkeys(self, stop_condition):
+    def _traverse_subkeys(self, stop_condition, collect_all_matching=False):
         """Perform simple BFS of subkeys, returning the key
         that successfully triggers the stop condition.
         Args:
             stop_condition: lambda or function pointer that takes a single argument
                             a key and returns a boolean value based on that key
+            collect_all_matching: boolean value, if True, the traversal collects and returns
+                            all keys meeting stop condition. If false, once stop
+                            condition is met, the key that triggered the condition '
+                            is returned.
         Return:
             the key if stop_condition is triggered, or None if not
         """
+        collection = []
         if not self._valid_reg_check():
             raise RegistryError("Cannot query values from invalid key %s" % self.key)
         with self.invalid_reg_ref_error_handler():
             queue = self.reg.subkeys
             for key in queue:
                 if stop_condition(key):
-                    return key
+                    if collect_all_matching:
+                        collection.append(key)
+                    else:
+                        return key
                 queue.extend(key.subkeys)
-            return None
+            return collection if collection else None
+
+    def _find_subkey_s(self, direct_subkey=None, search_key=None, collect_found=False):
+        """If non recursive, this method is the same as get subkey with error handling
+        Otherwise perform a BFS of subkeys until a key matching subkey name regex is found
+        Returns None or the first RegistryKey object corresponding to requested key name
+
+        Args:
+            direct_subkey (str): string representing subkey to be searched for.
+                                 Cannot be provided alongside `search_key`.
+            search_key (str): regex string represeting a subkey name structure
+                              to be matched against.
+                              Cannot be provided alongside `direct_subkey`
+            collect_found (bool): No-op if `direct_subkey` is specified
+        Return:
+            the desired subkey as a RegistryKey object, or none
+        """
+        if direct_subkey and search_key:
+            raise RuntimeError("Cannot specify both direct_subkey and search_key")
+        elif not (direct_subkey or search_key):
+            raise RuntimeError("One of direct_subkey/search_key required")
+        if direct_subkey:
+            return self.get_subkey(direct_subkey)
+        else:
+            return self._traverse_subkeys(search_key, collect_found=collect_found)
```

Comment:
I think this would be a non trivial effort to fully mock something as complex as the Windows registry. We'd need to both fake the `winreg` calls and create a fake registry object. I could, alternatively, ad-hoc patch a few our our reg interface calls to return expected values if that works for you?

Reply:
To be clear, I'm also totally fine with doing the former, I just wanted to be sure that's the direction we want to take.

Reply:
I'm inclined to say that monkeypatching the `winreg` calls could be difficult to track. That being said, there are quite a few, so perhaps as a follow-up refactoring PR? It looks like it would be straightforward (collect all the `winreg.` calls together and work from there) but let me know if I'm overlooking some complexity here.